### PR TITLE
Fix HPX pix_to_idx handling of non-spatial pix and add regression test

### PR DESCRIPTION
**Pull request by yaochengchen**
Description
This pull request fixes an inconsistency in the HPX geometry handling of pixel coordinates and adds a regression test to guard against it.
Previously, HpxGeom.pix_to_idx() always interpreted the first element of pix as a global HEALPix spatial index. This caused incorrect -1 indices (or downstream errors) when get_image_by_pix() passed only non-spatial pixel coordinates, which is the documented and expected usage of that method. The issue primarily affected partial-sky HPX geometries, where such global indices are almost never valid.
The fix updates HpxGeom.pix_to_idx() to explicitly distinguish between two valid input cases:
pix contains only non-spatial axes (as used by get_image_by_pix)
pix contains full spatial + non-spatial coordinates
For the non-spatial-only case, the method now delegates to axes.pix_to_idx() instead of interpreting the first element as a HEALPix spatial index. This preserves the existing behavior for full pixel inputs while making the HPX and WCS implementations consistent for non-spatial access.
A regression test is added to ensure that get_image_by_pix() works correctly for HPX maps when only non-spatial pixel coordinates are provided.
Dear reviewer
This PR is ready for review.
The change is intentionally limited to HPX-specific code and does not modify shared Map or get_image_by_pix() logic.
Existing behavior for full (spatial + non-spatial) pixel inputs is preserved.
A regression test is included and fails with the previous implementation while passing with this fix.
No documentation changes are included, as the behavior now matches the existing docstring and WCS behavior.
Feedback on the approach or edge cases is very welcome.

### CODE REVIEW

File: gammapy/maps/tests/test_core.py [hunk 1]
```diff
@@ -1124,3 +1124,30 @@ def test_quantity():
...
+        axes=axes,
+        unit="",
+    )
+    m.data = np.arange(m.data.size, dtype=float).reshape(m.data.shape)
+
+    # get_image_by_pix contract: pix contains ONLY non-spatial axes coordinates
+    pix = (1.2345, 0.1234)[: len(m.geom.axes)]
+
+    # old buggy HPXGeom.pix_to_idx mis-parses pix[0] as ipix -> -1 / IndexError
+    m_image = m.get_image_by_pix(pix)  # should NOT raise after the fix
```
Comment (QRemy) [inline gammapy/maps/tests/test_core.py]:
```suggestion
    m_image = m.get_image_by_pix(pix)
```

File: gammapy/maps/tests/test_core.py [hunk 2]
```diff
@@ -1124,3 +1124,30 @@ def test_quantity():
...
+        binsz=0.1,
+        width=10.0,
+        map_type="hpx",
+        skydir=SkyCoord(0.0, 30.0, unit="deg"),
+        axes=axes,
+        unit="",
+    )
+    m.data = np.arange(m.data.size, dtype=float).reshape(m.data.shape)
+
+    # get_image_by_pix contract: pix contains ONLY non-spatial axes coordinates
```
Comment (QRemy) [inline gammapy/maps/tests/test_core.py]:
```suggestion
```

Comment (yaochengchen) [reply gammapy/maps/tests/test_core.py]: Thanks @QRemy , Sorry for the messy coding, I will follow the coding format next time.

**Review by QRemy — APPROVED** [review]
Thanks @yaochengchen, no further comment from my side.

File: gammapy/maps/tests/test_core.py [hunk 3]
```diff
@@ -1124,3 +1124,28 @@ def test_quantity():
     m = Map.from_geom(geom)
     with pytest.raises(TypeError):
         m.data = 0 * u.deg
+
+
+@pytest.mark.parametrize("axes", [map_axes[:1], map_axes])
+def test_hpx_get_image_by_pix_non_spatial_only_regression(axes):
+    pytest.importorskip("healpy")
```
Comment (registerrier) [inline gammapy/maps/tests/test_core.py]:
```suggestion
```
the importorskip is applied to the whole file.

File: gammapy/maps/hpx/geom.py [hunk 4]
```diff
@@ -314,6 +314,13 @@ def pix_to_coord(self, pix):
 
     def pix_to_idx(self, pix, clip=False):
         # FIXME: Look for better method to clip HPX indices
+
+        # If pix only contains non-spatial coordinates (e.g. Map.get_image_by_pix),
+        # do NOT interpret pix[0] as HEALPix ipix.
+        if len(pix) == len(self.axes):
+            pix = pix_tuple_to_idx(pix)
```
Comment (registerrier) [inline gammapy/maps/hpx/geom.py:321]: Why is this step necessary? Why don't you pass `pix` directly. You are transforming it to idx twice.

Comment (registerrier) [reply gammapy/maps/hpx/geom.py:321]: I suppose this might be because `MapAxis.pix_to_idx` is in fact not really returning an `idx`...

File: gammapy/maps/tests/test_core.py [hunk 5]
```diff
@@ -1124,3 +1124,28 @@ def test_quantity():
     m = Map.from_geom(geom)
     with pytest.raises(TypeError):
         m.data = 0 * u.deg
+
+
+@pytest.mark.parametrize("axes", [map_axes[:1], map_axes])
+def test_hpx_get_image_by_pix_non_spatial_only_regression(axes):
```
Comment (registerrier) [inline gammapy/maps/tests/test_core.py]: I ma not sure you need to reimplement a test. You could already adapt existing ones as this test is re-using most of `test_map_get_image_by_pix`.

File: gammapy/maps/hpx/geom.py (hunk 4 above)
Comment (yaochengchen) [reply gammapy/maps/hpx/geom.py:321]: The extra pix_tuple_to_idx() step is intentional.
In this branch pix is a tuple of non-spatial pixel coordinates, while MapAxis.pix_to_idx() does not operate on tuple-style pix input but expects array-like indices for clipping/masking.
So this is not a double pix→idx conversion, but a tuple→array normalization followed by the axis-level index handling.
I agree the naming of pix_to_idx is a bit misleading here, which is likely the source of the confusion.

File: gammapy/maps/tests/test_core.py (hunk 3 above)
Comment (yaochengchen) [reply gammapy/maps/tests/test_core.py]: Thanks for pointing this out, I will modify this.

File: gammapy/maps/tests/test_core.py (hunk 5 above)
Comment (yaochengchen) [reply gammapy/maps/tests/test_core.py]: Thanks @registerrier, updated as suggested. Reusing test_map_get_image_by_pix with non-trivial data makes the coverage more straightforward.

**Comment by yaochengchen** [comment]
@registerrier @QRemy Sorry I accidently push re-review.

**Comment by QRemy** [comment]
The behaviour of `MapAxis.pix_to_idx()` could be clarified in a separate PR, merging this now.
